ublox/src/parser.rs

```diff
@@ -1,19 +1,172 @@
+pub struct Proto31;
 
-use crate::{
-    error::ParserError,
-    ubx_packets::{PacketRef, RTCM_SYNC_CHAR, SYNC_CHAR_1, SYNC_CHAR_2},
-};
+#[cfg(feature = "ubx_proto31")]
+impl UbxProtocol for Proto31 {
+    type PacketRef<'a> = crate::proto31::PacketRef<'a>;
+    const MAX_PAYLOAD_LEN: usize =
+        crate::ubx_packets::packets::packetref_proto31::MAX_PAYLOAD_LEN as usize;
+
+    fn match_packet(
+        class_id: u8,
+        msg_id: u8,
+        payload: &[u8],
+    ) -> Result<Self::PacketRef<'_>, ParserError> {
+        crate::ubx_packets::packets::packetref_proto31::match_packet(class_id, msg_id, payload)
+    }
+}
+
+/// Creates a parser for UBX Protocol v14 with a `Vec<u8>` buffer.
+///
+/// Requires the `ubx_proto14` and `alloc` (or `std`) features.
+#[cfg(all(feature = "ubx_proto14", any(feature = "std", feature = "alloc")))]
+pub fn proto14() -> Parser<Vec<u8>, Proto17> {
+    Parser::new(Vec::new())
+}
+
+/// Creates a parser for UBX Protocol v14 with a fixed-size buffer.
+///
+/// Requires the `ubx_proto14` feature.
+#[cfg(feature = "ubx_proto14")]
+pub fn proto14_with_buffer(buf: &mut [u8]) -> Parser<FixedLinearBuffer<'_>, Proto17> {
+    Parser::new(FixedLinearBuffer::new(buf))
+}
+
+/// Creates a parser for UBX Protocol v23 with a `Vec<u8>` buffer.
+///
+/// Requires the `ubx_proto23` and `alloc` (or `std`) features.
+#[cfg(all(feature = "ubx_proto23", any(feature = "std", feature = "alloc")))]
+pub fn proto23() -> Parser<Vec<u8>, Proto23> {
+    Parser::new(Vec::new())
+}
+
+/// Creates a parser for UBX Protocol v23 with a fixed-size buffer.
+///
+/// Requires the `ubx_proto23` feature.
+#[cfg(feature = "ubx_proto23")]
+pub fn proto23_with_buffer(buf: &mut [u8]) -> Parser<FixedLinearBuffer<'_>, Proto23> {
+    Parser::new(FixedLinearBuffer::new(buf))
+}
+
+/// Creates a parser for UBX Protocol v27 with a `Vec<u8>` buffer.
+///
+/// Requires the `ubx_proto27` and `alloc` (or `std`) features.
+#[cfg(all(feature = "ubx_proto27", any(feature = "std", feature = "alloc")))]
+pub fn proto27() -> Parser<Vec<u8>, Proto27> {
+    Parser::new(Vec::new())
+}
+
+/// Creates a parser for UBX Protocol v27 with a fixed-size buffer.
+///
+/// Requires the `ubx_proto27` feature.
+#[cfg(feature = "ubx_proto27")]
+pub fn proto27_with_buffer(buf: &mut [u8]) -> Parser<FixedLinearBuffer<'_>, Proto27> {
+    Parser::new(FixedLinearBuffer::new(buf))
+}
+
+/// Creates a parser for UBX Protocol v31 with a `Vec<u8>` buffer.
+///
+/// Requires the `ubx_proto31` and `alloc` (or `std`) features.
+#[cfg(all(feature = "ubx_proto31", any(feature = "std", feature = "alloc")))]
+pub fn proto31() -> Parser<Vec<u8>, Proto31> {
+    Parser::new(Vec::new())
+}
+
+/// Creates a parser for UBX Protocol v31 with a fixed-size buffer.
+///
+/// Requires the `ubx_proto31` feature.
+#[cfg(feature = "ubx_proto31")]
+pub fn proto31_with_buffer(buf: &mut [u8]) -> Parser<FixedLinearBuffer<'_>, Proto31> {
+    Parser::new(FixedLinearBuffer::new(buf))
+}
```

Comment:
This should probably just be deleted. We can write a parser builder if needed, but it's probably overkill.